[FLINK-14095][doc] Add Pravega connector in Flink Document

crazyzhou opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

 This pull request adds *Pravega connector* into the document.

## Brief change log

  - Add Pravega connector link in the connectors `index.md` and `ganrantee.md` page.
  - Add `pravega.md` page in the connector folder

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

wangxiyuan committed:
[FLINK-14241][test] Add arm support for docker test

Docker-compose doesn't have arm release bin file. Use pip3 install
for arm arch

This closes #9782

JingsongLi committed:
[FLINK-14543][table] Support partition for temporary table

crazyzhou committed:
[FLINK-14095] Add Pravega in Flink Document

Signed-off-by: Brian Zhou <brian.zhou@emc.com>

fpj commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,398 @@
+[Pravega](http://pravega.io/) is an open-source, high-performance, durable, elastic and strongly-consistent
+stream storage for continuous and unbounded data.
+
+Here are some need-to-know Pravega terminologies.
```
I suggest to replace this terminology list with the following paragraphs:

```
Pravega builds on the concept of *Stream*. A stream comprises one or more sequences of bytes. 
The unit of a stream is the *segment*. A *segment* is an unbounded sequence of bytes. A stream
can have one or more segments at any time, and the set of segments is allowed to change
dynamically with auto-scaling. Streams are grouped into scopes to avoid name collisions across
applications. 

One of the main APIs that Pravega provides is the event one. Events are serialized into bytes and
appended to segments. Pravega uses *routing keys* to map events to segments, they are
essentially sharded across open segments, and Pravega clients learn about open segments from
the control plane of Pravega.

Applications, including Flink jobs, read from a Pravega stream using reader groups. The readers
in a group share the workload of segments and coordinate the assignment of segments using a
Pravega feature that enables the coordination of state called *state synchronizer*. That coordination
is transparent to the application so that application does not have to handle dynamic changes to
the set of segments. This coordination also enables the generation of *checkpoints*, which are
essential to enable exactly-once semantics end-to-end with Flink. 

The foundation of checkpoints is an abstraction called *stream cut*: a collection of offsets spanning one or more segments. Checkpointing constitutes one of the main use cases of stream cuts, but stream cuts can additionally be used to set a position to begin reading or a position to end reading in a stream. To be able to resume or bound the computation over a stream of events, the Pravega API enables readers to obtain *stream cut* objects independent of checkpoints. 
```

crazyzhou replied:
Replaced.

fpj commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,398 @@
+- **Streamcut**  A `StreamCut` represents a specific position in a Pravega `Stream`, which may be obtained from various API interactions with the Pravega client.
+
+## Pravega connector
+Pravega connector can be used to build end-to-end real-time stream processing pipelines
```
We should say what the connector provides: source and sink implementations. As stated, it is too abstract. Here is a suggestion:

`The Pravega connector comprises source and sink implementations that enable jobs to use Pravega as an input of stream data and to store the output of jobs, respectively. The examples we show below in this document illustrate how an application sets it up to use the Pravega source and sink. See also the [Pravega Samples](https://github.com/pravega/pravega-samples) repository for more examples.`

crazyzhou replied:
Replaced

fpj commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,398 @@
+and Apache Flink for computing engine over the streams.
+
+### Features & Highlights
+- Exactly-once processing guarantees for both Reader and Writer, supporting end-to-end exactly-once processing pipelines
```
I'd remove the first part of this sentence and leave only: `Supports end-to-end exactly once semantics with checkpoints and transactions.`

crazyzhou replied:
Replaced.

fpj commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,398 @@
+- Table API support to access Pravega Streams for both Batch and Streaming use case.
+
+### Dependency
+To use this connector, please pick the proper dependency and add it to your project:
```
Let's make explicit references to maven and gradlew, and adapt the samples below accordingly:

`To use the connector, add the appropriate dependency to your project via Maven or Gradlew`.

crazyzhou replied:
I think as a Flink document, it should follow the Flink rule. Flink document only reference maven format for all its dependencies.
(e.g. https://ci.apache.org/projects/flink/flink-docs-stable/dev/connectors/kafka.html#usage 
https://ci.apache.org/projects/flink/flink-docs-release-1.9/getting-started/tutorials/datastream_api.html) 

There is not a template in the doc for switch in maven and other compiling tools, so I can only adopt your idea to keep showing these two together on the page. Your suggestion makes sense but I guess the Flink community wants to keep the tradition among the docs.

derekm replied:
Should we update this section for the 0.7 release?

derekm replied:
> There is not a template in the doc for switch in maven and other compiling tools, so I can only adopt your idea to keep showing these two together on the page. Your suggestion makes sense but I guess the Flink community wants to keep the tradition among the docs.

Here is a Flink quickstart document that shows both Maven and Gradle configs: https://ci.apache.org/projects/flink/flink-docs-stable/dev/projectsetup/java_api_quickstart.html#gradle

If it is just a short example, I think it is safe and helpful to include Gradle configuration stanzas here as well.

crazyzhou replied:
Makes sense, I'll add the `gradle` section.

fpj reviewed (changes requested):
The structure looks good, but it needs quite a bit of polishing. I have left some initial set of comments, but it needs more polishing than that.

xintongsong committed:
[hotfix][k8s] Remove duplicated taskManagerMemoryMB from KubernetesTaskManagerParameters.

xintongsong committed:
[FLINK-16439][k8s] KubernetesResourceManager starts workers with resources requested by SlotManager.

This means KubernetesResourceManager no longer:
- be aware of the default task executor resources
- assumes all workers are identical

This closes #11323.

xintongsong committed:
[FLINK-16439][runtime] Introduce PendingWorkerCounter for counting pending workers per WorkerResourceSpec in ActiveResourceManager.

xintongsong committed:
[hotfix][yarn][test] Avoid using Mockito for Container in YarnResourceManagerTest.

xintongsong committed:
[hotfix][yarn][test] Avoid using Mockito for ContainerStatus in YarnResourceManagerTest.

xintongsong committed:
[hotfix][yarn][test] Avoid using Mockito for AMRMClientAsync in YarnResourceManagerTest.

xintongsong committed:
[hotfix][yarn][test] Avoid using Mockito for NMClientAsync in YarnResourceManagerTest.

xintongsong committed:
[hotfix][yarn][test] Avoid using Mockito for TaskExecutorGateway in YarnResourceManagerTest.

xintongsong committed:
[FLINK-16438][runtime] Remove unused TaskExecutorProcessSpec from ActiveResourceManager.

xintongsong committed:
[hotfix][runtime] Code clean-ups in ResourceManagerRuntimeServices.

xintongsong committed:
[hotfix][runtime] Code deduplication in ResourceManagerFactory and its implementations.

xintongsong committed:
[FLINK-16440][runtime][metrics] Introduce SlotManagerMetricGroup for registering related metrics inside SlotManager.

xintongsong committed:
[FLINK-16440][core] Introduce multiplication for Resource.

xintongsong committed:
[FLINK-16440][runtime] Introduce multiplication for ResourceProfile.

xintongsong committed:
[FLINK-16440][runtime] Extend ResourceOverview and TaskManager(Details)Info for registered/free resources.

This closes #11363.

Commit:
[FLINK-16404][runtime] Avoid caching buffers for blocked input channels before barrier alignment

This commit is the first part of implementation to solve the dead lock problem when reducing the exclusive buffer of receiver side to 0.

Reducing the number of exclusive buffers of receiver side to 0 can bring several advantages (may at the cost of some performance regression). One is that memory can be saved from the reduced network buffer usage. Another important benefit is that the in-flight data can be reduced so we can speed up checkpoint in cases of back pressure. However, for the current implementation, reducing the exclusive buffer of receiver side can incur deadlock problem because all the floating buffers might be requested away by some blocked input channels and never recycled until barrier alignment.

To solve the problem, this commit mainly makes the following changes:
1. At sender side, after sending a checkpoint barrier when aligned exactly-once checkpoint mode is used, the outgoing channel will be blocked and no data will be sent out until the channel is unblocked.
2. At receiver side, no buffer will be stored in BufferStorage any more and after a checkpoint is completed or canceled, the receiver side will resume data consumption and unblock the upstream by sending a special event to the sender side.

Note that after this patch we still can't set the exclusive buffer of receiver side to 0 because there is still deadlock problem which will be totally solved in the following up patches.

zenfenan committed:
[FLINK-10114] Add ORC BulkWriter support for StreamingFileSink

This closes #11474.

becketqin committed:
[FLINK-15101][connector/common] Add the SourceCoordinator implementation

xintongsong committed:
[hotfix][runtime] Code clean-ups in BashJavaUtils and its test cases.

xintongsong committed:
[FLINK-16742][dist] config.sh does not assume BashJavaUtils always return two lines of results.

KarmaGYZ committed:
[FLINK-16605][core][config] Add slotmanager.max-number-of-slots config option

Commit:
[FLINK-16645][network] Limit the maximum backlogs in subpartitions

In the case of data skew, most of the buffers in #LocalBufferPool can be requested away by one
certain partition, which increases the in-flight data and slow down the buffer alignment.

To solve this problem, we add a config to limit the maximum backlog per subpartition. In this
way, we can check subpartition's backlog and make the stream task unavailable if the
limit is exceeded.

twalthr committed:
[FLINK-16779][table] Fix RAW type creation during type inference

twalthr committed:
[hotfix][sql-parser] Remove unused class SqlTimeUnit

twalthr committed:
[hotfix][table] Print more details during sink type mismatches

twalthr committed:
[FLINK-16779][table] Add RAW type support in DDL and functions

This fully introduces the RAW type in DDL and functions. It allows
to support opaque data types in parser and planner.

This closes #11568.

xintongsong committed:
[FLINK-16472][dist] Remove unused memory configuration logics from config.sh.

derekm commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,418 @@
+`0.6.1` is the Pravega version, you can find the latest release on
+the [GitHub Releases page](https://github.com/pravega/flink-connectors/releases).
+
+Since Pravega 0.6.0, the connector support multiple Flink versions.
```
```suggestion
Since Pravega 0.6.0, the connector supports multiple Flink versions.
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,418 @@
+the [GitHub Releases page](https://github.com/pravega/flink-connectors/releases).
+
+Since Pravega 0.6.0, the connector support multiple Flink versions.
+Pravega connector maintains the compatibility for recent *three* minor version of Flink.
```
```suggestion
Pravega connector maintains compatibility for the *three* most recent minor versions of Flink.
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.md:
```diff
@@ -0,0 +1,418 @@
+
+Since Pravega 0.6.0, the connector support multiple Flink versions.
+Pravega connector maintains the compatibility for recent *three* minor version of Flink.
+See below table for the supported versions.
```
```suggestion
See the table below for supported versions.
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+</div>
+
+### Serialization
+Flink's serialization/deserialization interfaces(eg. `SimpleStringSchema`) can be directly used.
```
```suggestion
Flink's serialization/deserialization interfaces(eg. `SimpleStringSchema`) can be used directly.
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+## Pravega Reader
+
+### Streaming API
+This connector provides a `FlinkPravegaReader` class to consume events from one or multiple Pravega stream in real-time.
```
```suggestion
This connector provides a `FlinkPravegaReader` class to consume events from one or multiple Pravega streams in real-time.
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+
+### Streaming API
+This connector provides a `FlinkPravegaReader` class to consume events from one or multiple Pravega stream in real-time.
+See the below for an example:
```
```suggestion
See below for an example:
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+
+### Batch API
+Similarly, a `FlinkPravegaInputFormat` class is offered for batch processing. It reads events of a stream as a `DataSet`.
+See the below for an example:
```
```suggestion
See below for an example:
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+</div>
+
+Details on more functionality such as checkpointing, optional start, metrics and
+watermark assigning can be found in the [document](http://pravega.io/docs/latest/connectors/connectors/index.html)
```
broken URL, poorly named link and missing punctuation

Is this the correct link? http://pravega.io/docs/latest/connectors/flink-connector/streaming/

Maybe link name is "Flink Streaming Connector document"?

crazyzhou replied:
I have corrected and merged this link into the end of the introduction as a conclusion.

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+
+### Streaming API
+This connector provides a `FlinkPravegaWriter` class to write events into a Pravega stream in real-time.
+See the below for an example:
```
```suggestion
See below for an example:
```

crazyzhou replied:
Fixed

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+2. **At-least-once** - All events are persisted in Pravega. Duplicate events
+are possible, due to retries or in case of failure and subsequent recovery.
+3. **Exactly-once** - All events are persisted in Pravega using a transactional approach integrated with the Flink checkpointing feature.
+For more details, please refer to [end-to-end exactly once blog](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html)
```
```suggestion
For more details, please refer to this [end-to-end exactly once blog](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html).
```

crazyzhou replied:
Fixed as
 ```
please refer to this [blog](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html) introducing end-to-end exactly once.
```

derekm commented on docs/dev/connectors/pravega.zh.md:
```diff
@@ -0,0 +1,418 @@
+</div>
+
+Details on more functionality such as event time ordering, watermark propagation and metrics can be found in the
+[Document](http://pravega.io/docs/latest/connectors/connectors/index.html)
```
broken URL, poorly named link and missing punctuation

crazyzhou replied:
Fixed

aljoscha committed:
[FLINK-17009] Move Java lambda and Scala Extension doc under streaming section

rmetzger committed:
[FLINK-16423][e2e] Introduce timeouts for HA tests

Before this change, HA tests were waiting indefinetly for certain conditions to be met.
This change introduces a timeout in the test scripts that kills the test and prints the Flink logs for debugging.

This closes #11831

aljoscha committed:
[FLINK-17009] Copy What is a Collection and Anatomy from api_concepts.md to datastream_api.md

This also modernizes the text in some places and replaces references to
"Collection" by "DataStream".

nielsbasjes committed:
[FLINK-16871][runtime] Make more build time information available at runtime

azagrebin committed:
[FLINK-16745][clients] Parse JobManagerProcessSpec from Configuration into ClusterSpecification

azagrebin committed:
[FLINK-16745][yarn] Start Yarn JM with FLIP-116 JVM memory args

azagrebin committed:
[FLINK-16745][k8s] Start Kubernetes JM with FLIP-116 JVM memory args

This closes #11675.

azagrebin committed:
[FLINK-16745][coordination] Remove unused container cutoff

xintongsong committed:
[hotfix][k8s][test] Allowing custom slot manager in KubernetesResourceManagerTest.

xintongsong committed:
[FLINK-16742][runtime] Ignore unknown command line options for BashJavaUtils.

xintongsong committed:
[FLINK-16742][dist] Extend and use BashJavaUtils for JM memory configuration in start-up scripts.

This closes #11545.

KarmaGYZ committed:
[FLINK-16605][runtime] Pass the slotmanager.max-number-of-slots to the SlotManagerImpl

xintongsong committed:
[FLINK-16438][yarn] Introduce WorkerSpecContainerResourceAdapter for converting between Flink WorkerResourceSpec and Yarn container Resource in YarnResourceManager.

xintongsong committed:
[FLINK-16438][yarn] YarnResourceManager starts workers with resources requested by SlotManager.

This means YarnResourceManager no longer:
- be aware of the default task executor resources
- assumes all workers are identical

tillrohrmann committed:
[hotfix] Add TimeUtils.toDuration

tillrohrmann committed:
[hotfix] Simplify AkkaRpcService.connectInternal by relying on Akka APIs

tillrohrmann committed:
[FLINK-17072] Add test case for random actor resolution

tillrohrmann committed:
[hotfix] Remove unused jobManagerAddress from StandaloneHAServices

tillrohrmann committed:
[hotfix] Remove unused RESOURCE_MANAGER_RPC_ENDPOINT_NAME from StandaloneHaServices

tillrohrmann committed:
[hotfix] Fix checkstyle violations in StandaloneHaServices

tillrohrmann committed:
[FLINK-17072] Let ResourceManager use random endpoint id

This commit ensures that every ResourceManager started by the same JVM process
will have a different endpoint id. This ensures that no new RM will receive
messages sent to an old RM.

tillrohrmann committed:
[FLINK-17072] Let Dispatcher use random endpoint id

This commit ensures that every Dispatcher, running in the same JVM,
is started with a different endpoint id. This ensures that no new
Dispatcher instance will receive messages sent to an old instance.

This closes #11693.

aljoscha committed:
[FLINK-17009] Copy Lazy Evaluation from api_concepts.md to datastream_api.md

This also slightly modifies/updates the text and removes the section
header.

aljoscha committed:
[FLINK-17009] Add intro to KeyedDataStream in state.md

aljoscha committed:
[FLINK-17009] Copy Specifying Keys from api_concepts.md to state.md

We also add a small introduction.

aljoscha committed:
[FLINK-17009] Remove tuple/expression keys from state.md

We now only document `KeySelector` user functions and discourage people
from using tuple/expression keys. See
https://issues.apache.org/jira/browse/FLINK-17074 for the reasoning
behind this.

aljoscha committed:
[FLINK-17009] Copy UDF section from api_concepts.md to toplevel DataStream section

This page is a bit tricky because we have a toplevel Java/Scala switcher
because the sections are actually different for the different APIs. We
therefore also don't have a table of contents.

aljoscha committed:
[FLINK-17009] Move Supported Data Type from api_concepts.md to types_serialization.md

This seems like a better fit, plus we want to get rid of the
api_concepts.md page.

aljoscha committed:
[FLINK-17009] Copy Accumulators & Counters from api_concepts.md to user_defined_functions.md

oliveryunchang committed:
[hotfix][docs] fix some typos in elasticsearch connector

xintongsong committed:
[hotfix][yarn] Code clean-ups in RegisterApplicationMasterResponseReflectorTest.

xintongsong committed:
[FLINK-16438][yarn] Decide WorkerSpecContainerResourceAdapter.MatchingStrategy based on RegisterApplicationMasterResponse.

This closes #11353.

KarmaGYZ committed:
[hotfix][runtime] Add sanity check to SlotManagerConfiguration

Myasuka committed:
[FLINK-14652][checkpointing] Refactor notifyCheckpointComplete to SubtaskCheckpointCoordinator

This closes #11721.

aljoscha committed:
[FLINK-17009] Remove api_concepts.md

aljoscha committed:
[FLINK-17009] Copy Specifying Keys section to DataSet documentation

We want to remove api_concepts.md so we need to preserve this part in
the DataSet documentation.

azagrebin committed:
[FLINK-16746] Add default 'jobmanager.memory.process.size: 1600m' to flink-conf.yaml

The resulting default JVM heap size for standalone deployment is 1024Mb as before FLIP-116.
The requested container size (Yarn and Kubernetes) will increase by default from 1024Mb to 1600Mb.

xintongsong committed:
[FLINK-16742][runtime] Refactor BashJavaUtils, move to separated package and extract configuration loading logics.

azagrebin committed:
[hotfix] Remove 'TaskManager' from MemoryBackwardsCompatibilityUtils#getLegacyTaskManagerHeapMemoryIfExplicitlyConfigured

azagrebin committed:
[FLINK-16746][conf] Deprecate jobmanager.heap.size and expose new memory options

Deprecate legacy heap options: jobmanager.heap.size (and update deprecation of jobmanager.heap.mb)
to use jobmanager.memory.heap.size for standalone
and jobmanager.memory.process.size for containerized (Yarn/K8s) environments.

Expose the new JM memory options of FLIP-116 in user docs.

This closes #11787.

bowenli86 committed:
[FLINK-17175][core] StringUtils.arrayToString() should consider Object[] lastly

bowenli86 committed:
[FLINK-16812][jdbc] support array types in PostgresRowConverter

closes #11766

kl0u committed:
[FLINK-16661] Move the ClassPathPackagedProgramRetriever to flink-clients

This is done because the ClassPathPackagedProgramRetriever is going to be
also used by other ApplicationEntrypoints in the future, including the
YarnApplicationClusterEntryPoint.

kl0u committed:
[FLINK-16661] Add support for Application Mode to Yarn

This commit adds Yarn support for Application Mode (FLIP-85).
Shipping of the user-jar is done the same way the
'yarn.per-job-cluster.include-user-jar' flag does it.

kl0u committed:
[FLINK-16661][cli] Add the 'run-application' command

This command in the CLI aims at supporting execution of
user applications in Application Mode (see FLIP-85) from
the CLI. It goes through the ExecutorCLI and NOT through
the FlinkYarnSessionCLI and the user has to write sth like:

./bin/flink run-application
  -t yarn-application
  -Djobmanager.memory.process.size=2048m
  -Dtaskmanager.memory.process.size=4096m
  ../examples/streaming/StateMachineExample.jar

kl0u committed:
[FLINK-16661][yarn] Add the YarnApplicationClusterEntryPoint

This is the ClusterEntryPoint to be used when running an
application on Yarn, in "Application Mode" (see in FLIP-85).

echauchot committed:
[FLINK-17074] Deprecate java Datastream.keyBy methods that do not use KeySelector

kl0u committed:
[FLINK-16661] Consolidating code in ApplicationClusterEntrypoints

xintongsong committed:
[hotfix][runtime][k8s] Renaming methods for better code readability.

wuchong committed:
[FLINK-17169][core] Add shortString() method to RowKind to get short string representation

This closes #11797

wuchong committed:
[FLINK-17169][table-blink] Refactor BaseRow to use RowKind instead of byte header

This closes #11797

zlb1028 committed:
[FLINK-16662] Let PackagedProgramUtils create JobGraph under the user classloader

Currently, in web submission JobGraph creation happens using the system classloader.
This can lead to failures when, for example, the JobGraph creation uses code generation
and needs the user code. This PR fixes it by doing the Jobgraph generation under the
user classloader.

bowenli86 committed:
[FLINK-16473][doc][jdbc] add documentation for JDBCCatalog and PostgresCatalog

closes #11804

wuchong committed:
[FLINK-17169][table-planner-blink] Rename StreamRecordUtils#record() to insertRecord()

This also Renames StreamRecordUtils#retractRecord() to updateBeforeRecord().

This closes #11797

Commit:
[hotfix][checkpointing] Make CheckpointBarrierUnaligner#processEndOfPartition always return false

For unaligned checkpoint, no buffer is stored in BufferStorage so CheckpointBarrierUnaligner#processEndOfPartition should always return false.

This closes #11351.

Commit:
[hotfix] Add more information to CheckpointCoordinatorConfiguration#toString

This closes #11351.

link3280 committed:
[FLINK-17403][test] Fix invalid classpath in BashJavaUtilsITCase

This closes #11934.

alpinegizmo committed:
[FLINK-17237][docs] Add Intro to DataStream API tutorial

This closes #11834.

alpinegizmo committed:
[FLINK-17236][docs] Add Tutorials section overview

This closes #11826.

wuchong committed:
[FLINK-17169][table-blink] Fix allowLateness shouldn't affect producing updates of emit strategy

This closes #11797

GJL committed:
[FLINK-17180][runtime] Implement SchedulingPipelinedRegion interface

Implement interfaces

  - Toplogy#getAllPipelinedRegions()
  - Topology#getPipelinedRegionOfVertex(ExecutionVertexID)

in DefaultExecutionTopology to enable retrieval of pipelined regions.

alpinegizmo committed:
[hotfix][docs] change KeySelector in program dataflow figure from String to lambda

alpinegizmo committed:
[FLINK-17238][docs] Data Pipelines and ETL tutorial

This closes #11842.

alpinegizmo committed:
[FLINK-17239][docs] Streaming Analytics tutorial

This closes #11843.

azagrebin committed:
[hotfix] Fix ProcessMemoryUtilsTestBase#testExceptionShouldContainRequiredConfigOptions to run also for TM

GJL committed:
[FLINK-17180][runtime] Use SchedulingPipelinedRegion in RegionPartitionReleaseStrategy

Avoid re-computing pipelined regions in the
RegionPartitionReleaseStrategy using the PipelinedRegionComputeUtil.
Instead, rely on the pipelined regions provided by the Topology.

GJL committed:
[hotfix][runtime] Throw exception from TestingSchedulingPipelinedRegion#getVertex() for unknown vertices

GJL committed:
[FLINK-17180][runtime] Use SchedulingPipelinedRegion in RestartPipelinedRegionFailoverStrategy

Avoid re-computing pipelined regions in the
RestartPipelinedRegionFailoverStrategy using the
PipelinedRegionComputeUtil. Instead, rely on the pipelined regions
provided by the Topology.

This closes #11857.

kl0u committed:
[FLINK-16661] Allow to set ApplicationStatus when shutting down cluster

Currently, when calling DispatcherGateway#shutdownCluster(), the
ApplicationStatus is set to SUCCESS automatically. This is not sufficient
when using the method to shut down the cluster due to a problem. In
addition, this results in frameworks like Yarn displaying the application
status as succeeded in any case. This commit adds a method to the
dispatcherGateway where the Application status can be specified and
uses the method in the ApplicationDispatcherBootstrap#stop() which is
used to shut down the cluster in application mode. We did not add the
new method to the RestfulGateway because it does not fit there.

alpinegizmo committed:
[FLINK-17240][docs] Add tutorial on Event-driven Apps

This closes #11845.

alpinegizmo committed:
[FLINK-17241][docs] Add Fault Tolerance tutorial

This closes #11846.

aljoscha committed:
[minor] Simplify PlanExecutor.execute() signature

becketqin committed:
[FLINk-17268][connector/common] Fix the bug of missing records during SplitFetcher wakeup. Added unit test for SplitFetcher.

zhuzhurk committed:
[FLINK-17021][table-planner-blink] Blink batch planner set GlobalDataExchangeMode

becketqin committed:
[hotfix] Remove the unused Boundedness parameter.

KarmaGYZ committed:
[FLINK-13639] Refactor the IntermediateResultPartitionID to consist of IntermediateDataSetID and partitionIndex

azagrebin committed:
[hotfix][mesos] Deduplicate code in mesos JM start bash scripts

azagrebin committed:
[FLINK-17048][mesos] Add memory related JVM args to Mesos JM startup scripts

This closes #11863.

becketqin committed:
[FLINK-17268][connector/common] Fix the shutdown sequence in SplitFetcher to avoid blocking on shutdown.
Fix flaky SourceReaderTestBase#testAvailableOnEmptyQueue().

kl0u committed:
[minor] Remove empty check in ConfigUtils#encodeCollectionToConfig()

tillrohrmann committed:
[hotfix] Remove unused TOTAL_FLINK_MEMORY_MB field

tillrohrmann committed:
[FLINK-17301][tests] Use proper AkkaRpcService for TaskManagerRunnerStartupTest

The problem is that we used Mockito to mock a RpcService. Due to this, some of the
fields which are later used are being null.

JingsongLi committed:
[FLINK-15901][hive] Support partitioned generic table in HiveCatalog

tillrohrmann committed:
Revert "[FLINK-14499][metric] MetricRegistry#getMetricQueryServiceGatewayRpcAddress is Nonnull"

This reverts commit 0cf9c2846ad1b293b91059d172dd68dd0ffbf803.

tillrohrmann committed:
[hotfix] Fix checkstyle violations in TaskExecutorPartitionLifecycleTest

tillrohrmann committed:
[hotfix] Let TaskExecutorOperatorEventHandlingTest extend TestLogger

tillrohrmann committed:
[FLINK-17301][tests] Set BIND_HOST to localhost in TaskManagerRunnerStartupTest

We need to configure BIND_HOST to localhost, otherwise Netty will use the
wildcard addresss to bind to an IPv6 address whose port is not blocked.

This closes #11864.

SteNicholas committed:
[FLINK-16766][python] Support creating StreamTableEnvironment without passing StreamExecutionEnvironment (#11769)

rkhachatryan committed:
[FLINK-17323][task] Return NO_MORE_DATA from ChannelStateReader for unknown channels

This closes #11861.

lirui-apache committed:
[FLINK-17138][sql-client] Close catalogs when close the session


This closes #11756

bowenli86 committed:
[FLINK-17333][doc] add doc for 'create catalog' ddl

closes #11871

lirui-apache committed:
[FLINK-17209][table] Allow users to specify dialect in sql-client yaml


This close #11788

WeiZhong94 committed:
[FLINK-16669][python] Introduce PythonFunctionFactory to provide a way to get the meta information of a Python UDF used in Java

This closes #11749.

WeiZhong94 committed:
[FLINK-16669][python][table] Support parse and register Python UDF via sql ddl

tillrohrmann committed:
[hotfix] Remove unused docs.rest.NoOpTransientBlobService

tillrohrmann committed:
[hotfix] Make NoOpTransientBlobService a public testing class

tillrohrmann committed:
[FLINK-17308] Pass ScheduledExecutorService to WebMonitorEndpoint

HuangXingBo committed:
[FLINK-17125][python] Add a Usage Notes Page to Answer Common Questions Encountered by PyFlink Users

This closes #11878

twalthr committed:
[FLINK-17373][table] Support the NULL type for function calls

Enables support for untyped NULL literals in function calls. The null
type remains a helper type. We don't expose this type through the DDL
or allow it as a function return type.

More cases of NULL can be supported once we enable type coercion.

This closes #11904.

tillrohrmann committed:
[FLINK-17308] Pass in ExecutionGraphCache to WebMonitorEndpoint

tillrohrmann committed:
[FLINK-17308] Extract ExecutionGraphCache interface and rename impl into DefaultExecutionGraphCache

This commit extracts the ExecutionGraphCache interface from the implementation and renames
the latter into DefaultExecutionGraphCache. Moreover, it introduces the NoOpExecutionGraphCache
implementation which is used for the DocumentingDispatcherRestEndpoint.

tillrohrmann committed:
[FLINK-17308] Add regular cleanup task for ExecutionGraphCache

The WebMonitorEndpoint now schedules are regular cleanup task which
runs every 2 * WebOptions.REFRESH_INTERVAL and tries to clean up
expired ExecutionGraphCache entries. This ensures that we will remove
unused entries.

This closes #11879.

rmetzger committed:
[hotfix] Set kafka-schema-registry-client to 4.1.0 in NOTICE

molsionmo committed:
[FLINK-17226][metrics][prometheus] Remove relocations

molsionmo committed:
[FLINK-17227][metrics][datadog] Remove relocations

echauchot committed:
[FLINK-17074] Deprecate java Datastream.partitionCustom methods that do not use KeySelector

aljoscha committed:
[FLINK-17338] Increase timeouts on LocalExecutorITCase

These tests are very heavy because they're repeatedly compiling SQL
queries. This test takes unacceptably long, as a mitigation this
increases the timeouts to try and make it more stable on CI.

alpinegizmo committed:
[FLINK-17352][docs] Fix doc links using site.baseurl w/ link tag

This closes #11885

KurtYoung committed:
[FLINK-17339][table-api] Change default planner to blink

This closes #11910

KurtYoung committed:
[FLINK-17339][table-planner] Update test cases due to default planner change.

zjffdu committed:
[FLINK-10911][scala-shell] Enable flink-scala-shell with Scala 2.12

KurtYoung committed:
[FLINK-17339][sql-client] Change default planner to blink.

KurtYoung committed:
[FLINK-17339][table-planner-blink] Simplify test cases due to default planner changing.

KurtYoung committed:
[FLINK-17339][examples] Update examples due to default planner changing.

KurtYoung committed:
[FLINK-17339][misc] Update tests due to default planner changing.

dianfu committed:
[FLINK-16485][python] Support vectorized Python UDF in batch mode of old planner (#11859)

sjwiesman committed:
[hotfix] fix broken links in documentation

kl0u committed:
[FLINK-16661] Move the static job id setting to ApplicationDispatcherBootstrap

HA support and static Job Ids go hand-in-hand, as HA requires that
the id of a job graph stays fixed across consecutive executions. In
addition, no 2 jobs can have the same job id while executing on the
same cluster. This commit consolidates this logic (in the context
of Application Mode) in one place, the ApplicationDispatcherBootstrap.

tillrohrmann committed:
[FLINK-14816] Let TaskManagerThreadDumpHandler return JSON response

Instead of serving plain text, this commit changes the TaskManagerThreadDumpHandler
to return a JSON response. This allows to further extend this handler to not
only generate a thread dump of all threads but also for a sub set. Morever,
it allows to return a more structured return value in the future.

This closes #11887.

GJL committed:
[hotfix][filesystems] Remove unused StopWatch

premsantosh committed:
[FLINK-17359] Fix EntropyInjector to resolve entropy for ClassLoaderFixingFileSystem

HuangXingBo committed:
[FLINK-17188][python] Use pip instead of conda to install flake8 and sphinx (#11836)

HuangXingBo committed:
[FLINK-15971][python] Adjust the default value of bundle size and bundle time (#11881)

Commit:
[FLINK-17344][test] Fix unstability on IdleTime metric test.

Previously issue was that a thread could finish before the main thread reached .wait() call

lirui-apache committed:
[FLINK-16802][hive] Set schema info in JobConf for Hive readers


This closes #11735

zhuzhurk committed:
[FLINK-17020][runtime] Introduce GlobalDataExchangeMode for JobGraph generation

alpinegizmo committed:
[hotfix][docs] link tags in zh docs require .zh.md links

WeiZhong94 committed:
[FLINK-16667][python][client] Support the Python dependency configuration options in CliFrontend (#11702)

twalthr committed:
[FLINK-17370][table-common] Ensure deterministic type inference extraction

This ensures that function signatures are always extracted in a deterministic
way. Esp. when it comes to implicit casting, a deterministic type inference
is important to evaluate sequences like f(INT) || f(BIGINT).

This closes #11902.

WeiZhong94 committed:
[FLINK-16971][python][sql-client] Support Python UDF in SQL Client.

This closes #11901.

rmetzger committed:
[FLINK-17374][travis] Remove travis-related files

This closes #11921

rmetzger committed:
[FLINK-17374][travis] Remove tools/travis directory

rmetzger committed:
[FLINK-17374][travis] Further removals of travis-mentions

WeiZhong94 committed:
[FLINK-17345][python][table] Support register and get Python UDF in Catalog (#11884)

aljoscha committed:
[FLINK-16661] Add YARN application end-to-end test

This renames the existing test_yarn_kerberos_docker.sh to
test_yarn_job_kerberos_docker.sh to reflect that it is a per-job test
and adds a new test based on that for YARN application mode.

KurtYoung committed:
[FLINK-16935][table-planner-blink] Enable or delete most of the ignored test cases in blink planner.

This closes #11874

JingsongLi committed:
[FLINK-14258][table][filesystem] Integrate file system connector to streaming sink


This closes #11796

JingsongLi committed:
[FLINK-15006][table-planner] Add option to shuffle-by-partition when dynamic inserting


This closes #11853

aljoscha committed:
Partial Revert "[FLINK-15669] [sql-client] Fix SQL client can't cancel flink job"

This partially reverts commit 0650e4b6b6a9725c3a9589a110b202478a68f90c.
Only the tests are reverted because the fix is still good.  The added
tests don't work because of a race condition: if the query finishes
before the test cancels them the job status will be FINISHED, and not
CANCELLED, as the text expects.

fpompermaier committed:
[FLINK-17385][jdbc][postgres] Handled problem of numeric with 0 precision

closes #11914

xintongsong committed:
[hotfix][yarn] Code clean-ups in WorkerSpecContainerResourceAdapter.

HuangXingBo committed:
[FLINK-17311][python] Change the version of cython to 0.29.16

HuangXingBo committed:
[FLINK-17311][AZP] Add the logic of compressed in tgz before uploading artifacts

This closes #11860

HuangXingBo committed:
[FLINK-17395][release][python] Add the sign and shasum logic for PyFlink wheel packages

This closes #11915.

xintongsong committed:
[FLINK-17390][yarn] Fix Hadoop 2.10+ compatibility for WorkerSpecContainerResourceAdapter.

This closes #11916.

docete committed:
[FLINK-15591][sql-parser] Support parsing TEMPORARY in table definition

docete committed:
[FLINK-15591][table] Support create/drop temporary table in both planners

This closes #11917

echauchot committed:
[FLINK-17074] Deprecate scala Datastream.keyBy methods that do not use KeySelector

echauchot committed:
[FLINK-17074] Deprecate scala Datastream.partitionCustom methods that do not use KeySelector

zjuwangg committed:
[FLINK-17313] Fix type validation error when sink table ddl contains columns with precision of decimal/varchar (#11848)

aljoscha committed:
Revert "[FLINK-10911][scala-shell] Enable flink-scala-shell with Scala 2.12"

This reverts commit b7ce6119158346504e0ddd8d0f110859304eacbd.

Our per-push CI does not cover Scala 2.12 so we didn't notice that
ScalaShellITCase does not actually pass.

KarmaGYZ committed:
[FLINK-16965][metrics][graphite] Add GraphiteReporterFactory

Commit:
[FLINK-17315][tests] Increased timeouts of UnalignedCheckpointITCase.

caozhen1937 committed:
[hotfix][docs] Fix typo

HuangXingBo committed:
[FLINK-17414][python][docs] Update the documentation about PyFlink build about Cython support

This closes #11927.

aljoscha committed:
[hotfix] Remove site.baseurl from {% link %} tags

link tags already include site.baseurl.

KarmaGYZ committed:
[FLINK-16605][runtime] Make the SlotManager respect the max limitation for slots

This closes #11615.

KarmaGYZ committed:
[hotfix][tests] Deduplicate code in SlotManagerImplTest

lirui-apache committed:
[FLINK-17210][sql-parser][hive] Implement database DDLs for Hive dialect


This closes #11791

HuangXingBo committed:
[FLINK-17423][python] Support Java UDTF and UDAF in Blink planner under batch mode for PyFlink

This closes #11932.

lamberken committed:
[hotfix] Add comments to FileType enum values

lamberken committed:
[hotfix] Factor out putTransientBlobStream from requestFileuploadFilePath

lamberken committed:
[FLINK-14816] Add thread dump feature for taskmanager

docete committed:
[FLINK-17111][table] Support SHOW VIEWS in Flink SQL

This closes #11869

alpinegizmo committed:
[FLINK-17432][docs-training] Rename Tutorials to Training for better SEO (#11931)

[FLINK-17432][docs-training] Rename Tutorials to Training for better SEO

kl0u committed:
[FLINK-16661] Report the correct ApplicationStatus when job cancelled

kl0u committed:
[minor] Move getUsrLibDir() to YarnEntrypointUtils

zhijiangW committed:
[hotfix][network] Rename ResultPartitionWriter#initializeState to #readRecoveredState

zhijiangW committed:
[hotfix][checkpointing] Use practical ChannelStateReader instead of NO_OP

z8r-coder committed:
[FLINK-17334][hive] Flink does not support HIVE UDFs with primitive return types


This closes #11876

wuchong committed:
[FLINK-16996][table-common] Add necessary methods to internal data structures

- Add to primitive array methods to ArrayData
- Add get(ArrayData, int, LogicalType) utility to ArrayData
- Add get(Object key) to GenericMapData
- Add get(RowData, int, LogicalType) utility to RowData
- Add toString() to RowData

This closes #11925

wuchong committed:
[FLINK-16996][table-common] Add binary implementations of internal data structures

This closes #11925

zhijiangW committed:
[FLINK-17440][network] Resolve potential buffer leak in output unspilling for unaligned checkpoints

If there are any exceptions thrown while interacting with ChannelStateReader#readOutputData, we should close the
respective BufferConsumer to avoid leak.

wuchong committed:
[FLINK-16996][table-runtime-blink] Implement all the data structures and serializers around RowData

This closes #11925

wuchong committed:
[FLINK-16996][table-runtime-blink] Remove legacy data formats (BaseRow)

This closes #11925

wuchong committed:
[FLINK-16996][table] Refactor planner and runtime to use new data structures

This closes #11925

wuchong committed:
[FLINK-16996][python] Refactor pyflink to use new data structures

This closes #11925

wuchong committed:
[FLINK-16996][parquet] Refactor parquet connector to use new data structures

This closes #11925

wuchong committed:
[FLINK-16996][orc] Refactor ORC connector to use new data structures

This closes #11925

wuchong committed:
[FLINK-16996][hive] Refactor Hive connector to use new data structures

This closes #11925

JingsongLi committed:
[FLINK-17391][filesystem] sink.rolling-policy.time.interval default value should be bigger


This closes #11933

JingsongLi committed:
[FLINK-17455][table][filesystem] Move FileSystemFormatFactory to table common


This closes #11941

shuiqiangchen committed:
[FLINK-17254][python][docs] Improve the PyFlink documentation and examples to use SQL DDL for source/sink definition.

This closes #11945.

zhengcanbin committed:
[FLINK-17332][k8s] Fix restart policy not equals to Never for native task manager pods

This closes #11949 .

dianfu committed:
[FLINK-17148][python] Support converting pandas DataFrame to Flink Table (#11832)

JingsongLi committed:
[FLINK-17325][orc] Integrate orc to file system connector


This closes #11942

Myasuka committed:
[FLINK-13390][checkpointing] Clarify the exact meaning of state size when executing incremental checkpoint

This closes #9289.

twalthr committed:
[hotfix][table-common] Improve terminology around catalog table options

twalthr committed:
[hotfix][table-common] Enable receiving isBounded in DynamicTableSink

GJL committed:
[FLINK-17473][tests] Remove unused test utilities

Remove classes:
  - ArchivedExecutionBuilder
  - ArchivedExecutionJobVertexBuilder
  - ArchivedExecutionVertexBuilder

This closes #11958.

twalthr committed:
[FLINK-16997][table-common] Add new factory interfaces and discovery utilities

Implements the new factory interfaces mentioned in FLIP-95. Adds new factory utilities
that can be used for FLIP-122 and future factories.

It adds TestDynamicTableFactory and TestFormatFactory for reference implementations
of new factories.

This closes #11959.

wuchong committed:
[FLINK-17461][formats][json] Support JSON serialization and deseriazation schema for RowData type

This closes #11944

aljoscha committed:
[minor] Verify cluster shutdown status in ApplicationDispatcherBootstrap

This also verifies cluster shutdown status in other tests

zhengcanbin committed:
[hotfix][javadocs] Fix typo

twalthr committed:
[hotfix][table-common] Reduce number of public extraction classes

WeiZhong94 committed:
[FLINK-17471][python] Move the LICENSE and NOTICE file to the package root of the PyFlink source distribution. (#11956)

WeiZhong94 committed:
[FLINK-17436][python] Fix the IllegalAccessError caused by package-private access control when submitting Python job via "flink run". (#11930)

kl0u committed:
[FLINK-16661][cli] Fix log configuration forwarding for yarn application mode

carp84 committed:
[FLINK-16901][legal] Correctly handle the THIRD_PARTY_NOTICES file in kinesis connector bundled dependency

This closes #11964.

Commit:
[FLINK-17315][tests] Temporary disable UnalignedCheckpointITCase.

Apparently, it's causing deadlocks.

sjwiesman committed:
[hotfix] Fix formatting on python common questions

twalthr committed:
[FLINK-17489][core] Support any kind of array in StringUtils.arrayAwareToString()

This closes #11969.

sjwiesman committed:
[hotfix] fix donwload link

carp84 committed:
[FLINK-17483][legal] Update flink-sql-connector-elasticsearch7 NOTICE file to correctly reflect bundled dependencies

tweise committed:
[FLINK-17496][kinesis] Upgrade amazon-kinesis-producer to 0.14.0

Gmoonnight committed:
[FLINK-16103][docs-zh] Translate "Configuration" page of "Table API & SQL" into Chinese

This closes #11909

sjwiesman committed:
[FLINK-17499][state-processor-api] LazyTimerService used to register timers via State Processing API incorectly mixes event time timers with processing time timers

This closes #11980

kl0u committed:
[FLINK-17515][yarn] Move YARN staging functionality to a separate class

This closes #11991.

GJL committed:
[FLINK-17501][qs] Improve logging in AbstractServerHandler#channelRead(ChannelHandlerContext, Object)

Log errors as soon as possible, i.e., before serializing and sending an
error response to the client. This mitigates the problem that errors
might be masked by other errors that are caused by the error handling
logic.

This closes #11984.

dawidwys committed:
[FLINK-16034][table-api, docs] Updat documentation with new Java Expression DSL (#11533)

dawidwys committed:
[FLINK-16804] Deprecate all Table methods that accept expressions as a
String

All methods of Table (and related classes like e.g. GroupedTable) that
expect expressions as a String were deprecated in favor of the
Expression DSL counterparts. This commit also removes majority of the
usages of the deprecated methods. The remaining methods explicitly test
the behavior of parsing/converting the String DSL.

This closes #11989

tillrohrmann committed:
[hotfix][tests] Add TestingTaskManagerActions implementation

tillrohrmann committed:
[FLINK-17514] Let TaskCancelerWatchDog call TaskManagerActions.notifyFatalError with non-null cause

In order to avoid NPE we have to call TaskManagerActions.notifyFatalError with non-null arguments. This
commit changes the behaviour accordingly.

tillrohrmann committed:
[FLINK-17514] Fail fatally if the TaskCancelerWatchDog encounters exception in run method

If the TaskCancelerWatchDog encounters an exception in the run method, then we can no longer
guarantee that it will do its job. Hence, it is best to fail fatally by letting the exception
bubble up so that it is handled by the uncaught exception handler.

tillrohrmann committed:
[FLINK-17514] Harden ExceptionUtils.tryEnrichTaskManagerError, .isMetaspaceOutOfMemoryError and .isDirectOutOfMemoryError to handle null values

Error handling code should not fail. Hence, this commit changes the methods ExceptionUtils.tryEnrichTaskManagerError,
ExceptionUtils.isMetaspaceOutOfMemoryError and ExceptionUtils.isDirectOutOfMemoryError to properly handle a null argument.

This closes #11994.

leonardBang committed:
[FLINK-17420][table sql / api]Cannot alias Tuple and Row fields when converting DataStream to Table (#11951)11951)

alpinegizmo committed:
[FLINK-17244][docs] Update the Getting Started page (#11988)

* Update docs/getting-started/index.md

wuchong committed:
[FLINK-17462][format][csv] Support CSV serialization and deseriazation schema for RowData type

This closes #11962

shuiqiangchen committed:
[FLINK-17255][python] Support HBase connector descriptor in PyFlink. (#11955)

JingsongLi committed:
[FLINK-17460][orc][parquet] Create sql-jars for parquet and orc


This closes #11946

fpj reviewed (approved):
It looks good, when you propose it in the Apache Flink jira, I'll make another pass there.
